// react-image-annotate/MainLayout/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import React, { useState, useRef } from "react";
import Grid from "@material-ui/core/Grid";
import { makeStyles } from "@material-ui/core/styles";
import Sidebar from "../Sidebar";
import ImageCanvas from "../ImageCanvas";
import Header from "../Header";
import IconTools from "../IconTools";
import styles from "./styles";
import useKey from "use-key-hook";
import classnames from "classnames";
import { useSettings } from "../SettingsProvider";
import SettingsDialog from "../SettingsDialog";
import Fullscreen from "../Fullscreen";
import getActiveImage from "../Annotator/reducers/get-active-image";
import useImpliedVideoRegions from "./use-implied-video-regions";
var useStyles = makeStyles(styles);
export default (function (_ref) {
  var _Object$assign;

  var state = _ref.state,
      dispatch = _ref.dispatch,
      _ref$alwaysShowNextBu = _ref.alwaysShowNextButton,
      alwaysShowNextButton = _ref$alwaysShowNextBu === void 0 ? false : _ref$alwaysShowNextBu,
      _ref$alwaysShowPrevBu = _ref.alwaysShowPrevButton,
      alwaysShowPrevButton = _ref$alwaysShowPrevBu === void 0 ? false : _ref$alwaysShowPrevBu,
      RegionEditLabel = _ref.RegionEditLabel;
  var classes = useStyles();
  var settings = useSettings();
  var memoizedActionFns = useRef({});

  var action = function action(type) {
    for (var _len = arguments.length, params = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      params[_key - 1] = arguments[_key];
    }

    var fnKey = "".concat(type, "(").concat(params.join(","), ")");
    if (memoizedActionFns.current[fnKey]) return memoizedActionFns.current[fnKey];

    var fn = function fn() {
      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      return params.length > 0 ? dispatch(_objectSpread({
        type: type
      }, params.reduce(function (acc, p, i) {
        return acc[p] = args[i], acc;
      }, {}))) : dispatch(_objectSpread({
        type: type
      }, args[0]));
    };

    memoizedActionFns.current[fnKey] = fn;
    return fn;
  };

  var _getActiveImage = getActiveImage(state),
      currentImageIndex = _getActiveImage.currentImageIndex,
      activeImage = _getActiveImage.activeImage;

  var nextImage;

  if (currentImageIndex !== null) {
    nextImage = state.images[currentImageIndex + 1];
  }

  useKey(function () {
    return dispatch({
      type: "CANCEL"
    });
  }, {
    detectKeys: [27]
  });
  var isAVideoFrame = activeImage && activeImage.frameTime !== undefined;
  var impliedVideoRegions = useImpliedVideoRegions(state);
  return /*#__PURE__*/React.createElement(Fullscreen, {
    enabled: state.fullScreen,
    onChange: function onChange(open) {
      if (!open) {
        action("HEADER_BUTTON_CLICKED", "buttonName")("Exit Fullscreen");
      }
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: classnames(classes.container, state.fullScreen && "Fullscreen")
  }, /*#__PURE__*/React.createElement("div", {
    className: classes.headerContainer
  }, /*#__PURE__*/React.createElement(Header, {
    onHeaderButtonClick: action("HEADER_BUTTON_CLICKED", "buttonName"),
    videoMode: state.annotationType === "video",
    alwaysShowNextButton: alwaysShowNextButton,
    alwaysShowPrevButton: alwaysShowPrevButton,
    inFullScreen: state.fullScreen,
    isAVideoFrame: isAVideoFrame,
    nextVideoFrameHasRegions: !nextImage || nextImage.regions && nextImage.regions.length > 0,
    videoDuration: state.videoDuration,
    multipleImages: state.images && state.images.length > 1,
    title: state.annotationType === "image" ? activeImage ? activeImage.name : "No Image Selected" : state.videoName || "",
    onChangeCurrentTime: action("CHANGE_VIDEO_TIME", "newTime"),
    videoPlaying: state.videoPlaying,
    currentVideoTime: state.currentVideoTime,
    keyframes: state.keyframes
  })), /*#__PURE__*/React.createElement("div", {
    className: classes.workspace
  }, /*#__PURE__*/React.createElement("div", {
    className: classes.iconToolsContainer
  }, /*#__PURE__*/React.createElement(IconTools, {
    enabledTools: state.enabledTools,
    showTags: state.showTags,
    selectedTool: state.selectedTool,
    onClickTool: action("SELECT_TOOL", "selectedTool")
  })), /*#__PURE__*/React.createElement("div", {
    className: classes.imageCanvasContainer
  }, state.annotationType === "image" && !state.selectedImage ? /*#__PURE__*/React.createElement("div", {
    className: classes.noImageSelected
  }, "No Image Selected") : /*#__PURE__*/React.createElement("div", {
    style: {
      height: "100%",
      width: "100%"
    }
  }, /*#__PURE__*/React.createElement(ImageCanvas, Object.assign({}, settings, (_Object$assign = {
    key: state.selectedImage,
    showTags: state.showTags,
    allowedArea: state.allowedArea,
    regionClsList: state.regionClsList,
    regionTagList: state.regionTagList,
    regions: state.annotationType === "image" ? activeImage.regions || [] : impliedVideoRegions,
    realSize: activeImage ? activeImage.realSize : undefined,
    videoPlaying: state.videoPlaying,
    imageSrc: state.annotationType === "image" ? state.selectedImage : null,
    videoSrc: state.annotationType === "video" ? state.videoSrc : null,
    pointDistancePrecision: state.pointDistancePrecision,
    createWithPrimary: state.selectedTool.includes("create"),
    dragWithPrimary: state.selectedTool === "pan",
    zoomWithPrimary: state.selectedTool === "zoom",
    showPointDistances: state.showPointDistances
  }, _defineProperty(_Object$assign, "pointDistancePrecision", state.pointDistancePrecision), _defineProperty(_Object$assign, "videoTime", state.annotationType === "image" ? state.selectedImageFrameTime : state.currentVideoTime), _defineProperty(_Object$assign, "onMouseMove", action("MOUSE_MOVE")), _defineProperty(_Object$assign, "onMouseDown", action("MOUSE_DOWN")), _defineProperty(_Object$assign, "onMouseUp", action("MOUSE_UP")), _defineProperty(_Object$assign, "onChangeRegion", action("CHANGE_REGION", "region")), _defineProperty(_Object$assign, "onBeginRegionEdit", action("OPEN_REGION_EDITOR", "region")), _defineProperty(_Object$assign, "onCloseRegionEdit", action("CLOSE_REGION_EDITOR", "region")), _defineProperty(_Object$assign, "onDeleteRegion", action("DELETE_REGION", "region")), _defineProperty(_Object$assign, "onBeginBoxTransform", action("BEGIN_BOX_TRANSFORM", "box", "directions")), _defineProperty(_Object$assign, "onBeginMovePolygonPoint", action("BEGIN_MOVE_POLYGON_POINT", "polygon", "pointIndex")), _defineProperty(_Object$assign, "onAddPolygonPoint", action("ADD_POLYGON_POINT", "polygon", "point", "pointIndex")), _defineProperty(_Object$assign, "onSelectRegion", action("SELECT_REGION", "region")), _defineProperty(_Object$assign, "onBeginMovePoint", action("BEGIN_MOVE_POINT", "point")), _defineProperty(_Object$assign, "onImageLoaded", action("IMAGE_LOADED", "image")), _defineProperty(_Object$assign, "RegionEditLabel", RegionEditLabel), _defineProperty(_Object$assign, "onImageOrVideoLoaded", action("IMAGE_OR_VIDEO_LOADED", "metadata")), _defineProperty(_Object$assign, "onChangeVideoTime", action("CHANGE_VIDEO_TIME", "newTime")), _defineProperty(_Object$assign, "onChangeVideoPlaying", action("CHANGE_VIDEO_PLAYING", "isPlaying")), _Object$assign))))), /*#__PURE__*/React.createElement("div", {
    className: classes.sidebarContainer
  }, /*#__PURE__*/React.createElement(Sidebar, {
    debug: window.localStorage.$ANNOTATE_DEBUG_MODE && state,
    taskDescription: state.taskDescription,
    images: state.images,
    regions: activeImage ? activeImage.regions : null,
    history: state.history,
    currentImage: activeImage,
    labelImages: state.labelImages,
    imageClsList: state.imageClsList,
    imageTagList: state.imageTagList,
    keyframes: state.keyframes,
    currentVideoTime: state.currentVideoTime,
    onChangeImage: action("CHANGE_IMAGE", "delta"),
    onSelectRegion: action("SELECT_REGION", "region"),
    onDeleteRegion: action("DELETE_REGION", "region"),
    onSelectImage: action("SELECT_IMAGE", "image"),
    onChangeRegion: action("CHANGE_REGION", "region"),
    onRestoreHistory: action("RESTORE_HISTORY"),
    onChangeVideoTime: action("CHANGE_VIDEO_TIME", "newTime"),
    onDeleteKeyframe: action("DELETE_KEYFRAME", "time"),
    onShortcutActionDispatched: dispatch
  }))), /*#__PURE__*/React.createElement(SettingsDialog, {
    open: state.settingsOpen,
    onClose: function onClose() {
      return dispatch({
        type: "HEADER_BUTTON_CLICKED",
        buttonName: "Settings"
      });
    }
  })));
});
